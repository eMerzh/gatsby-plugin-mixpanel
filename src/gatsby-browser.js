import React from 'react'

import { MixpanelProvider, mixpanel } from '.'

const isEnable = options =>
  (process.env.NODE_ENV === `production` || options.enableOnDevMode) &&
  options.apiToken

const getOptions = pluginOptions => {
  const defaultsOptions = {
    apiToken: null,
    enableOnDevMode: true,
    debug: false,
  }
  const options = { ...defaultsOptions, ...pluginOptions }
  return { ...options, isEnable: isEnable(options) }
}

exports.onRouteUpdate = ({ location }, pluginOptions) => {
  const options = getOptions(pluginOptions)
  if (options.isEnable && options.pageViews) {
    const mixpanelEvent = options.pageViews[location.pathname]
    if (mixpanelEvent) {
      mixpanel.track(mixpanelEvent, location)
    }
  }
}

exports.onClientEntry = (skip, pluginOptions) => {
  const options = getOptions(pluginOptions)

  if (!options.isEnable) {
    mixpanel.init('disable', { autotrack: false })
    mixpanel.disable()
    return
  }

  mixpanel.init(options.apiToken, { debug: options.debug })
  console.log('ini')
}

exports.wrapRootElement = ({ element }) => (
  <MixpanelProvider>{element}</MixpanelProvider>
)
