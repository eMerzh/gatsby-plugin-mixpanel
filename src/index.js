import React, {
  PureComponent,
  createContext,
  cloneElement,
  useContext,
} from 'react'
import mixpanel from 'mixpanel-browser'

const MixpanelContext = createContext()

export const MixpanelProvider = ({ children }) => (
  <MixpanelContext.Provider value={mixpanel}>
    {children}
  </MixpanelContext.Provider>
)

export const MixpanelConsumer = ({ children }) => (
  <MixpanelContext.Consumer>
    {mixpanel => cloneElement(children, { mixpanel })}
  </MixpanelContext.Consumer>
)

export const withMixpanel = () => WrappedComponent => {
  return class extends PureComponent {
    render() {
      return (
        <MixpanelConsumer>
          <WrappedComponent {...this.props} />
        </MixpanelConsumer>
      )
    }
  }
}

export const useMixpanel = () => useContext(MixpanelContext)

export { mixpanel }
