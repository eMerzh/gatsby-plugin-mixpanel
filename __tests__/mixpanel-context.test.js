import React, { PureComponent } from 'react'
import { MixpanelConsumer, MixpanelProvider, withMixpanel } from '../src'
import { mount } from 'enzyme'

const DumbComponent = () => {
  return <h1>Dumb</h1>
}

@withMixpanel()
class DumbComponentWithDecoratorMixpanel extends PureComponent {
  render() {
    return <h1>Dumb</h1>
  }
}

const wrapMixpanelProvider = component => {
  return <MixpanelProvider>{component}</MixpanelProvider>
}

describe('Get mixpanel library', () => {
  it('on a consumer component', () => {
    const wrapper = mount(
      wrapMixpanelProvider(
        <MixpanelConsumer>
          <DumbComponent />
        </MixpanelConsumer>
      )
    )

    expect(wrapper.find('DumbComponent').props('mixpanel')).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('withMixpanel wrapped function', () => {
    const ComponentWithMixpanel = withMixpanel()(DumbComponent)
    const wrapper = mount(wrapMixpanelProvider(<ComponentWithMixpanel />))

    expect(wrapper.find('DumbComponent').props('mixpanel')).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('withMixpanel decorator function', () => {
    const wrapper = mount(
      wrapMixpanelProvider(<DumbComponentWithDecoratorMixpanel />)
    )

    expect(
      wrapper.find('DumbComponentWithDecoratorMixpanel').props('mixpanel')
    ).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })

})
