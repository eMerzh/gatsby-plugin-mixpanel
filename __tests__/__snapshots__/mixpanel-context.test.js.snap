// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mixpanel Consumer Consumer component 1`] = `
<MixpanelProvider>
  <MixpanelConsumer>
    <DumbComponent
      mixpanel={
        MixpanelLib {
          "_": Object {
            "HTTPBuildQuery": [Function],
            "JSONDecode": [Function],
            "JSONEncode": [Function],
            "UUID": [Function],
            "base64Encode": [Function],
            "bind": [Function],
            "bind_instance_methods": [Function],
            "cookie": Object {
              "get": [Function],
              "parse": [Function],
              "remove": [Function],
              "set": [Function],
              "set_seconds": [Function],
            },
            "dom_query": [Function],
            "each": [Function],
            "encodeDates": [Function],
            "escapeHTML": [Function],
            "extend": [Function],
            "formatDate": [Function],
            "getHashParam": [Function],
            "getQueryParam": [Function],
            "identity": [Function],
            "include": [Function],
            "includes": [Function],
            "info": Object {
              "browser": [Function],
              "browserVersion": [Function],
              "campaignParams": [Function],
              "device": [Function],
              "os": [Function],
              "pageviewInfo": [Function],
              "people_properties": [Function],
              "properties": [Function],
              "referringDomain": [Function],
              "searchEngine": [Function],
              "searchInfo": [Function],
            },
            "inherit": [Function],
            "isArguments": [Function],
            "isArray": [Function],
            "isBlockedUA": [Function],
            "isDate": [Function],
            "isElement": [Function],
            "isEmptyObject": [Function],
            "isFunction": [Function],
            "isNumber": [Function],
            "isObject": [Function],
            "isString": [Function],
            "isUndefined": [Function],
            "keys": [Function],
            "localStorage": Object {
              "error": [Function],
              "get": [Function],
              "is_supported": [Function],
              "parse": [Function],
              "remove": [Function],
              "set": [Function],
            },
            "register_event": [Function],
            "safewrap": [Function],
            "safewrap_class": [Function],
            "safewrap_instance_methods": [Function],
            "strip_empty_properties": [Function],
            "timestamp": [Function],
            "toArray": [Function],
            "trim": [Function],
            "truncate": [Function],
            "utf8Encode": [Function],
            "values": [Function],
          },
          "init": [Function],
        }
      }
    >
      <h1>
        Dumb
      </h1>
    </DumbComponent>
  </MixpanelConsumer>
</MixpanelProvider>
`;

exports[`Mixpanel Consumer withMixpanel decorator function 1`] = `
<MixpanelProvider>
  <_class>
    <MixpanelConsumer>
      <DumbComponentWithDecoratorMixpanel
        mixpanel={
          MixpanelLib {
            "_": Object {
              "HTTPBuildQuery": [Function],
              "JSONDecode": [Function],
              "JSONEncode": [Function],
              "UUID": [Function],
              "base64Encode": [Function],
              "bind": [Function],
              "bind_instance_methods": [Function],
              "cookie": Object {
                "get": [Function],
                "parse": [Function],
                "remove": [Function],
                "set": [Function],
                "set_seconds": [Function],
              },
              "dom_query": [Function],
              "each": [Function],
              "encodeDates": [Function],
              "escapeHTML": [Function],
              "extend": [Function],
              "formatDate": [Function],
              "getHashParam": [Function],
              "getQueryParam": [Function],
              "identity": [Function],
              "include": [Function],
              "includes": [Function],
              "info": Object {
                "browser": [Function],
                "browserVersion": [Function],
                "campaignParams": [Function],
                "device": [Function],
                "os": [Function],
                "pageviewInfo": [Function],
                "people_properties": [Function],
                "properties": [Function],
                "referringDomain": [Function],
                "searchEngine": [Function],
                "searchInfo": [Function],
              },
              "inherit": [Function],
              "isArguments": [Function],
              "isArray": [Function],
              "isBlockedUA": [Function],
              "isDate": [Function],
              "isElement": [Function],
              "isEmptyObject": [Function],
              "isFunction": [Function],
              "isNumber": [Function],
              "isObject": [Function],
              "isString": [Function],
              "isUndefined": [Function],
              "keys": [Function],
              "localStorage": Object {
                "error": [Function],
                "get": [Function],
                "is_supported": [Function],
                "parse": [Function],
                "remove": [Function],
                "set": [Function],
              },
              "register_event": [Function],
              "safewrap": [Function],
              "safewrap_class": [Function],
              "safewrap_instance_methods": [Function],
              "strip_empty_properties": [Function],
              "timestamp": [Function],
              "toArray": [Function],
              "trim": [Function],
              "truncate": [Function],
              "utf8Encode": [Function],
              "values": [Function],
            },
            "init": [Function],
          }
        }
      >
        <h1>
          Dumb
        </h1>
      </DumbComponentWithDecoratorMixpanel>
    </MixpanelConsumer>
  </_class>
</MixpanelProvider>
`;

exports[`Mixpanel Consumer withMixpanel wrapped function 1`] = `
<MixpanelProvider>
  <_class>
    <MixpanelConsumer>
      <DumbComponent
        mixpanel={
          MixpanelLib {
            "_": Object {
              "HTTPBuildQuery": [Function],
              "JSONDecode": [Function],
              "JSONEncode": [Function],
              "UUID": [Function],
              "base64Encode": [Function],
              "bind": [Function],
              "bind_instance_methods": [Function],
              "cookie": Object {
                "get": [Function],
                "parse": [Function],
                "remove": [Function],
                "set": [Function],
                "set_seconds": [Function],
              },
              "dom_query": [Function],
              "each": [Function],
              "encodeDates": [Function],
              "escapeHTML": [Function],
              "extend": [Function],
              "formatDate": [Function],
              "getHashParam": [Function],
              "getQueryParam": [Function],
              "identity": [Function],
              "include": [Function],
              "includes": [Function],
              "info": Object {
                "browser": [Function],
                "browserVersion": [Function],
                "campaignParams": [Function],
                "device": [Function],
                "os": [Function],
                "pageviewInfo": [Function],
                "people_properties": [Function],
                "properties": [Function],
                "referringDomain": [Function],
                "searchEngine": [Function],
                "searchInfo": [Function],
              },
              "inherit": [Function],
              "isArguments": [Function],
              "isArray": [Function],
              "isBlockedUA": [Function],
              "isDate": [Function],
              "isElement": [Function],
              "isEmptyObject": [Function],
              "isFunction": [Function],
              "isNumber": [Function],
              "isObject": [Function],
              "isString": [Function],
              "isUndefined": [Function],
              "keys": [Function],
              "localStorage": Object {
                "error": [Function],
                "get": [Function],
                "is_supported": [Function],
                "parse": [Function],
                "remove": [Function],
                "set": [Function],
              },
              "register_event": [Function],
              "safewrap": [Function],
              "safewrap_class": [Function],
              "safewrap_instance_methods": [Function],
              "strip_empty_properties": [Function],
              "timestamp": [Function],
              "toArray": [Function],
              "trim": [Function],
              "truncate": [Function],
              "utf8Encode": [Function],
              "values": [Function],
            },
            "init": [Function],
          }
        }
      >
        <h1>
          Dumb
        </h1>
      </DumbComponent>
    </MixpanelConsumer>
  </_class>
</MixpanelProvider>
`;
